addons/web/controllers/webmanifest.py

```diff
@@ -95,3 +100,117 @@ def offline(self):
+    def scoped_app_icon_png(self, app_id, add_padding=False):
+        """ Returns an app icon created with a fixed size in PNG. It is required for Safari PWAs """
+        # To begin, we take the first icon available for the app
+        app_icon = self._get_scoped_app_icons(app_id)[0]
+
+        if app_icon['type'] == "image/svg+xml":
+            # We don't handle SVG images here, let's look for the module icon if possible
+            manifest = modules.module.get_manifest(app_id)
+            add_padding = True
+            if len(manifest) > 0 and manifest['icon']:
+                icon_src = manifest['icon']
+            else:
+                icon_src = f"/{self._icon_path()}"
+        else:
+            icon_src = app_icon['src']
+            if not add_padding:
+                # A valid icon is explicitly provided, we can use it directly
+                return request.redirect(app_icon['src'])
+
+        # Now that we have the image source, we can generate a PNG image
+        img_io = self._generate_png_icon(icon_src, add_padding)
+        return send_file(img_io, request.httprequest.environ, mimetype='image/png')
+
+    @http.route('/web/manifest.scoped_app_manifest', type='http', auth='public', methods=['GET'])
+    def scoped_app_manifest(self, app_id, path, app_name=''):
+        """ Returns a WebManifest dedicated to the scope of the given app. A custom scope and start
+            url are set to make sure no other installed PWA can overlap the scope (e.g. /odoo)
+        """
+        _path = unquote(path)
+        _app_name = unquote(app_name) if app_name else self._get_scoped_app_name(app_id)
+        webmanifest = {
+            'icons': self._get_scoped_app_icons(app_id),
+            'name': _app_name,
+            'scope': _path,
+            'start_url': _path,
+            'display': 'standalone',
+            'background_color': '#714B67',
+            'theme_color': '#714B67',
+            'prefer_related_applications': False,
+            'shortcuts': self._get_scoped_app_shortcuts(app_id)
+        }
+        body = json.dumps(webmanifest, default=ustr)
+        response = request.make_response(body, [
+            ('Content-Type', 'application/manifest+json'),
+        ])
+        return response
+
+    def _get_scoped_app_shortcuts(self, app_id):
+        return []
+
+    def _get_scoped_app_name(self, app_id):
+        manifest = modules.module.get_manifest(app_id)
+        if len(manifest) > 0:
+            return manifest['name']
+
+    def _get_scoped_app_icons(self, app_id):
+        try:
+            file_path(f'{app_id}/static/description/icon.svg')
+            src = f'/{app_id}/static/description/icon.svg'
+        except FileNotFoundError:
+            src = f"/{self._icon_path()}"
+        return [{
+            'src': src,
+            'sizes': 'any',
+            'type': mimetypes.guess_type(src)[0] or 'image/png'
+        }]
+
+    def _generate_png_icon(self, icon_src, add_padding=False):
```

Comment:
can you verify such a tool doesn't exist already inside `odoo.tools.image`? Also this shouldn't be defined on the controller itself but in the tools and imported as a function here.

Reply:
Yes I'll rework that part